src/librustdoc/html/render/print_item.rs

```diff
@@ -1415,3 +1423,46 @@ fn document_non_exhaustive(w: &mut Buffer, item: &clean::Item) {
         w.write_str("</div>");
     }
 }
+
+fn document_type_layout(w: &mut Buffer, cx: &Context<'_>, ty_def_id: DefId) {
+    if !cx.shared.show_type_layout {
+        return;
+    }
+
+    let tcx = cx.tcx();
+    let param_env = tcx.param_env(ty_def_id);
+    let ty = tcx.type_of(ty_def_id);
+    match tcx.layout_of(param_env.and(ty)) {
+        Ok(ty_layout) => {
+            writeln!(w, "<h2 class=\"small-section-header\">Layout</h2>");
+            // FIXME: remove `no-toggle` (not just here but everywhere)
+            //        once toggling is handled using pure HTML & CSS (#83332)
+            writeln!(w, "<div class=\"docblock no-toggle\">");
+            writeln!(
+                w,
+                "<div class=\"warning\"><p><strong>Note:</strong> Most layout information is \
+                 completely unstable and may be different between compiler versions and platforms. \
+                 The only exception is types with certain <code>repr(...)</code> attributes. \
+                 Please see the Rust Reference’s \
+                 <a href=\"https://doc.rust-lang.org/reference/type-layout.html\">“Type Layout”</a> \
+                 chapter for details on type layout guarantees.</p></div>"
+            );
+            if ty_layout.layout.abi.is_unsized() {
+                writeln!(w, "<p><strong>Size:</strong> (unsized)</p>");
+            } else {
+                let bytes = ty_layout.layout.size.bytes();
+                writeln!(
+                    w,
+                    "<p><strong>Size:</strong> {size} byte{pl}</p>",
+                    size = bytes,
+                    pl = if bytes == 1 { "" } else { "s" },
+                );
+            }
+            writeln!(w, "</div>");
+        }
+        // Layout errors can occur with valid code, e.g. if you try to get the layout
+        // of a generic type such as `Vec<T>`. In case of a layout error, we just
+        // don't show any layout information.
+        Err(_) => {}
```

Comment:
One remaining thing I was thinking about is should we still show a layout section in case of a `LayoutError::Unknown`, but just say that we couldn't figure out the layout, possibly due to generics? Otherwise we may get bug reports saying that the layout section is not there.

Reply:
What about recursive types? :)

Reply:
Those shouldn't pass the compiler unless they have indirection, in which case the size will be known. I think we should only show this for correct code (that may use generics).

Reply:
Fine by me!